File: torch_nntile/csrc/models/llama.cpp
/*! @copyright (c) 2026-present Skolkovo Institute of Science and Technology
 *                              (Skoltech), Russia. All rights reserved.
 *
 * @file torch_nntile/csrc/models/llama.cpp
 * Llama causal LM - port of deleted ``nntile::model::llama``.
 */

#include <torch_nntile/models/llama.hh>

#include "nntile_gemm.h"
#include "nntile_rms_norm.h"
#include "nntile_rope.h"
#include "nntile_sdpa.h"
#include "nntile_model_transpose.h"

#include <cmath>
#include <stdexcept>
#include <vector>

namespace torch_nntile::models
{

namespace
{

torch::Tensor causal_mask_host(int64_t seq)
{
    auto opts = torch::TensorOptions()
        .dtype(torch::kBool)
        .device(torch::kCPU);
    auto q = torch::arange(
        seq,
        torch::TensorOptions().dtype(torch::kLong).device(torch::kCPU))
        .unsqueeze(1);
    auto k = torch::arange(
        seq,
        torch::TensorOptions().dtype(torch::kLong).device(torch::kCPU))
        .unsqueeze(0);
    return (k <= q).to(opts);
}

torch::Tensor rms_norm(
    torch::Tensor x,
    torch::Tensor weight,
    double eps)
{
    auto out_rstd = rms_norm_forward(
        x,
        /*normalized_shape=*/std::vector<int64_t>{x.size(-1)},
        weight,
        eps);
    return std::get<0>(out_rstd);
}

void rope_sin_cos_host(
    int64_t batch,
    int64_t seq,
    int64_t head_dim,
    double rope_theta,
    torch::Tensor &sin_out,
    torch::Tensor &cos_out)
{
    if (head_dim % 2 != 0)
    {
        throw std::invalid_argument("rope: head_dim must be even");
    }
    int64_t const half = head_dim / 2;
    std::vector<float> inv(static_cast<std::size_t>(half));
    for (int64_t i = 0; i < half; ++i)
    {
        double idx = static_cast<double>(2 * i);
        inv[static_cast<std::size_t>(i)] = static_cast<float>(
            1.0
            / std::pow(
                rope_theta,
                idx / static_cast<double>(head_dim)));
    }
    sin_out = torch::empty(
        {batch, seq, half},
        torch::TensorOptions()
            .dtype(torch::kFloat32)
            .device(torch::kCPU));
    cos_out = torch::empty_like(sin_out);
    auto sin_a = sin_out.accessor<float, 3>();
    auto cos_a = cos_out.accessor<float, 3>();
    for (int64_t b = 0; b < batch; ++b)
    {
        for (int64_t s = 0; s < seq; ++s)
        {
            for (int64_t h = 0; h < half; ++h)
            {
                double angle = static_cast<double>(s)
                    * static_cast<double>(
                        inv[static_cast<std::size_t>(h)]);
                sin_a[b][s][h] = static_cast<float>(std::sin(angle));
                cos_a[b][s][h] = static_cast<float>(std::cos(angle));
            }
        }
    }
}

torch::Tensor apply_rope(
    torch::Tensor x,
    torch::Tensor sin,
    torch::Tensor cos)
{
    // x layout: [n_heads, batch, seq, head_dim] (after model_transpose).
    int64_t const n_heads = x.size(0);
    if (sin.dim() == 3)
    {
        int64_t const b = sin.size(0);
        int64_t const s = sin.size(1);
        int64_t const half = sin.size(2);
        sin = sin.view({1, b, s, half}).repeat({n_heads, 1, 1, 1});
        cos = cos.view({1, b, s, half}).repeat({n_heads, 1, 1, 1});
    }
    return rope(sin, cos, x);
}

torch::Tensor repeat_kv_heads(torch::Tensor x, int64_t n_rep)
{
    // NNGraph scale_slice on axis 1 for GQA:
    // [n_kv, B, S, D] -> [n_kv, n_rep, B, S, D] via view+repeat.
    if (n_rep == 1)
    {
        return x;
    }
    int64_t const n_kv = x.size(0);
    int64_t const b = x.size(1);
    int64_t const s = x.size(2);
    int64_t const d = x.size(3);
    return x.view({n_kv, 1, b, s, d})
        .repeat({1, n_rep, 1, 1, 1});
}

} // namespace

// -- LlamaAttentionImpl ----------------------------------------------------

LlamaAttentionImpl::LlamaAttentionImpl(LlamaConfig const &cfg) :
    n_heads(cfg.num_attention_heads),
    n_kv(cfg.num_key_value_heads),
    head_size(cfg.hidden_size / cfg.num_attention_heads),
    n_rep(cfg.num_attention_heads / cfg.num_key_value_heads),
    use_gqa(cfg.num_key_value_heads < cfg.num_attention_heads)
{
    if (cfg.hidden_size % cfg.num_attention_heads != 0)
    {
        throw std::invalid_argument(
            "LlamaAttention: hidden_size % n_heads != 0");
    }
    if (cfg.num_attention_heads % cfg.num_key_value_heads != 0)
    {
        throw std::invalid_argument(
            "LlamaAttention: n_heads % n_kv != 0");
    }
    int64_t const h = cfg.hidden_size;
    int64_t const hs = head_size;
    if (use_gqa)
    {
        q_weight = register_parameter(
            "q_weight",
            torch::empty({h, hs, n_kv, n_rep}));
        o_weight = register_parameter(
            "o_weight",
            torch::empty({hs, n_kv, n_rep, h}));
    }
    else
    {
        q_weight = register_parameter(
            "q_weight",
            torch::empty({h, hs, n_heads}));
        o_weight = register_parameter(
            "o_weight",
            torch::empty({hs, n_heads, h}));
    }
    k_weight = register_parameter(
        "k_weight",
        torch::empty({h, hs, n_kv}));
    v_weight = register_parameter(
        "v_weight",
        torch::empty({h, hs, n_kv}));
    torch::nn::init::normal_(q_weight, 0.0, 0.02);
    torch::nn::init::normal_(k_weight, 0.0, 0.02);
    torch::nn::init::normal_(v_weight, 0.0, 0.02);
    torch::nn::init::normal_(o_weight, 0.0, 0.02);
}

torch::Tensor LlamaAttentionImpl::forward(
    torch::Tensor x,
    torch::Tensor const &sin,
    torch::Tensor const &cos,
    torch::Tensor const &mask)
{
    auto q = gemm(x, q_weight, /*ndim=*/1, /*batch_ndim=*/0);
    q = model_transpose(q, use_gqa ? 2 : 1);
    auto k = model_transpose(
        gemm(x, k_weight, /*ndim=*/1, /*batch_ndim=*/0),
        /*model_ndim=*/1);
    auto v = model_transpose(
        gemm(x, v_weight, /*ndim=*/1, /*batch_ndim=*/0),
        /*model_ndim=*/1);
    q = apply_rope(q, sin, cos);
    k = apply_rope(k, sin, cos);
    auto k_sdpa = k;
    auto v_sdpa = v;
    int64_t batch_ndim = 2;
    if (use_gqa)
    {
        k_sdpa = repeat_kv_heads(k, n_rep);
        v_sdpa = repeat_kv_heads(v, n_rep);
        batch_ndim = 3;
    }
    auto attn = sdpa_kernel(
        q,
        k_sdpa,
        v_sdpa,
        mask,
        batch_ndim);
    attn = model_transpose(attn, /*model_ndim=*/3);
    int64_t const out_ndim = use_gqa ? 3 : 2;
    return gemm(attn, o_weight, out_ndim, /*batch_ndim=*/0);
}

// -- LlamaMLPImpl ---------------------------------------------------------

LlamaMLPImpl::LlamaMLPImpl(LlamaConfig const &cfg)
{
    int64_t const h = cfg.hidden_size;
    int64_t const i = cfg.intermediate_size;
    gate_weight = register_parameter(
        "gate_weight",
        torch::empty({i, h}));
    up_weight = register_parameter(
        "up_weight",
        torch::empty({i, h}));
    down_weight = register_parameter(
        "down_weight",
        torch::empty({h, i}));
    torch::nn::init::normal_(gate_weight, 0.0, 0.02);
    torch::nn::init::normal_(up_weight, 0.0, 0.02);
    torch::nn::init::normal_(down_weight, 0.0, 0.02);
}

torch::Tensor LlamaMLPImpl::forward(torch::Tensor x)
{
    auto gate = gemm(
        x,
        gate_weight,
        /*ndim=*/1,
        /*batch_ndim=*/0,
        /*trans_a=*/false,
        /*trans_b=*/true);
    auto up = gemm(
        x,
        up_weight,
        /*ndim=*/1,
        /*batch_ndim=*/0,
        /*trans_a=*/false,
        /*trans_b=*/true);
    auto hidden = torch::silu(gate) * up;
    return gemm(
        hidden,
        down_weight,
        /*ndim=*/1,
        /*batch_ndim=*/0,
        /*trans_a=*/false,
        /*trans_b=*/true);
}

// -- LlamaDecoderImpl -----------------------------------------------------

LlamaDecoderImpl::LlamaDecoderImpl(LlamaConfig const &cfg) :
    rms_eps(cfg.rms_norm_eps)
{
    input_norm_w = register_parameter(
        "input_norm_w",
        torch::ones({cfg.hidden_size}));
    attn = register_module("attention", LlamaAttention(cfg));
    post_attn_norm_w = register_parameter(
        "post_attn_norm_w",
        torch::ones({cfg.hidden_size}));
    mlp = register_module("mlp", LlamaMLP(cfg));
}

torch::Tensor LlamaDecoderImpl::forward(
    torch::Tensor x,
    torch::Tensor const &sin,
    torch::Tensor const &cos,
    torch::Tensor const &mask)
{
    auto attn_out = attn->forward(
        rms_norm(x, input_norm_w, rms_eps),
        sin,
        cos,
        mask);
    auto post = x + attn_out;
    auto mlp_out = mlp->forward(
        rms_norm(post, post_attn_norm_w, rms_eps));
    return post + mlp_out;
}

// -- LlamaCausalImpl ------------------------------------------------------

LlamaCausalImpl::LlamaCausalImpl(LlamaConfig cfg) : config(std::move(cfg))
{
    if (config.hidden_size % config.num_attention_heads != 0)
    {
        throw std::invalid_argument(
            "LlamaCausal: hidden_size % n_heads != 0");
    }
    embed_tokens = register_module(
        "embed_tokens",
        torch::nn::Embedding(config.vocab_size, config.hidden_size));
    torch::nn::ModuleList list;
    for (int64_t i = 0; i < config.num_hidden_layers; ++i)
    {
        list->push_back(LlamaDecoder(config));
    }
    layers = register_module("layers", list);
    weight_rms = register_parameter(
        "weight_rms",
        torch::ones({config.hidden_size}));
    lm_weight = register_parameter(
        "lm_weight",
        torch::empty({config.vocab_size, config.hidden_size}));
    torch::nn::init::normal_(lm_weight, 0.0, 0.02);
}

void LlamaCausalImpl::warm_rope_cache(
    int64_t batch,
    int64_t seq,
    torch::Device device)
{
    int64_t const head_dim =
        config.hidden_size / config.num_attention_heads;
    torch::Tensor sin_h;
    torch::Tensor cos_h;
    rope_sin_cos_host(
        batch,
        seq,
        head_dim,
        config.rope_theta,
        sin_h,
        cos_h);
    auto mask = causal_mask_host(seq);
    if (!device.is_cpu())
    {
        sin_h = sin_h.to(device);
        cos_h = cos_h.to(device);
        mask = mask.to(device);
    }
    rope_sin_ = sin_h;
    rope_cos_ = cos_h;
    cached_mask_ = mask;
    rope_cache_batch_ = batch;
    rope_cache_seq_ = seq;
}

torch::Tensor LlamaCausalImpl::forward(torch::Tensor input_ids)
{
    int64_t const b = input_ids.size(0);
    int64_t const s = input_ids.size(1);
    if (!rope_sin_.defined() || rope_cache_batch_ != b
        || rope_cache_seq_ != s
        || rope_sin_.device() != input_ids.device())
    {
        warm_rope_cache(b, s, input_ids.device());
    }
    auto x = embed_tokens->forward(input_ids);
    for (auto &module : *layers)
    {
        x = module->as<LlamaDecoderImpl>()->forward(
            x,
            rope_sin_,
            rope_cos_,
            cached_mask_);
    }
    x = rms_norm(x, weight_rms, config.rms_norm_eps);
    return gemm(
        x,
        lm_weight,
        /*ndim=*/1,
        /*batch_ndim=*/0,
        /*trans_a=*/false,
        /*trans_b=*/true);
}

} // namespace torch_nntile::models

File: torch_nntile/torch_nntile/nn/model/llama.py
# @copyright (c) 2026-present Skolkovo Institute of Science and Technology
#                              (Skoltech), Russia. All rights reserved.
#
# @file torch_nntile/torch_nntile/nn/model/llama.py
# Llama causal LM for device="nntile".

"""Llama stack mirroring ``nntile::model::llama`` (RMSNorm, RoPE, SiLU MLP).

Forward / backward keep activations on ``device=nntile`` end-to-end.
RoPE uses ``arange(seq)`` for every batch row. ``sin``/``cos`` are
``[seq, head_dim // 2]``; extra leading modes of ``x`` (heads, GQA,
batch) are the kernel batch ``n``. There is no ``[B, S]`` position
table.
"""

from __future__ import annotations

from dataclasses import dataclass

import torch
import torch.nn as nn
from torch import Tensor

from torch_nntile import _C
from torch_nntile.gemm import gemm
from torch_nntile.nn import Embedding, SiLU
from torch_nntile.nn.functional import add, mul
from torch_nntile.nn.linear import NntileLinear, prepare_sdpa_mask
from torch_nntile.nn.model.gpt2_minimal import make_causal_sdpa_mask
from torch_nntile.nn.sdpa import nntile_model_transpose, sdpa_kernel
from torch_nntile.normalization import rms_norm

try:
    from torch_nntile.rope import rope, rope_sin_cos_from_position_ids
except ImportError:  # pragma: no cover - stub if rope.py missing
    def rope(sin: Tensor, cos: Tensor, x: Tensor) -> Tensor:
        return x

    def rope_sin_cos_from_position_ids(
        position_ids: Tensor,
        head_dim: int,
        *,
        rope_theta: float = 10000.0,
    ) -> tuple[Tensor, Tensor]:
        del rope_theta
        half = head_dim // 2
        shape = (*tuple(position_ids.shape), half)
        z = torch.zeros(
            shape, dtype=torch.float32, device=position_ids.device
        )
        return z, torch.ones_like(z)


def _repeat_kv(x: Tensor, n_rep: int) -> Tensor:
    """GQA KV expand via classic ``scale_slice``, not ``unsqueeze``+``repeat``.

    ``aten::unsqueeze`` is a view that keeps the 4-D storage TensorRef, so
    SDPA backward can hand ``rope_backward`` a 5-D node vs 4-D ``dx``.
    Matches C++ ``repeat_kv_heads`` / NNGraph ``scale_slice`` on axis 1.
    ``x`` is ``[H_kv, B, S, D]`` -> ``[H_kv, n_rep, B, S, D]``.
    """
    if n_rep == 1:
        return x
    if x.device.type == "nntile":
        return _C.scale_slice(x, 1, n_rep)
    return x.unsqueeze(1).repeat(1, n_rep, 1, 1, 1)


@dataclass
class LlamaConfig:
    vocab_size: int = 32000
    hidden_size: int = 4096
    intermediate_size: int = 11008
    num_hidden_layers: int = 32
    num_attention_heads: int = 32
    num_key_value_heads: int = 32
    max_position_embeddings: int = 2048
    head_dim: int = 128
    rms_norm_eps: float = 1e-6
    rope_theta: float = 10000.0
    attention_bias: bool = False
    mlp_bias: bool = False
    tie_word_embeddings: bool = False
    eos_token_id: int = 2
    bos_token_id: int = 1
    name: str = "llama"

    def __post_init__(self) -> None:
        if (
            self.num_attention_heads > 0
            and self.hidden_size % self.num_attention_heads == 0
        ):
            self.head_dim = self.hidden_size // self.num_attention_heads

    def validate(self) -> None:
        if self.hidden_size % self.num_attention_heads != 0:
            raise ValueError(
                "LlamaConfig: hidden_size must be divisible by "
                "num_attention_heads"
            )
        if self.num_attention_heads % self.num_key_value_heads != 0:
            raise ValueError(
                "LlamaConfig: num_attention_heads must be divisible by "
                "num_key_value_heads"
            )


class LlamaRMSNorm(nn.Module):
    def __init__(self, hidden_size: int, eps: float = 1e-6) -> None:
        super().__init__()
        self.weight = nn.Parameter(torch.ones(hidden_size))
        self.eps = eps
        self.normalized_shape = (hidden_size,)

    def forward(self, x: Tensor) -> Tensor:
        return rms_norm(x, self.normalized_shape, self.weight, self.eps)


class LlamaMLP(nn.Module):
    """Gated MLP: ``down(SiLU(gate(x)) * up(x))``."""

    def __init__(self, config: LlamaConfig) -> None:
        super().__init__()
        h = config.hidden_size
        i = config.intermediate_size
        bias = config.mlp_bias
        self.gate_proj = NntileLinear(h, i, bias=bias)
        self.up_proj = NntileLinear(h, i, bias=bias)
        self.down_proj = NntileLinear(i, h, bias=bias)
        self.act = SiLU()

    def forward(self, x: Tensor) -> Tensor:
        return self.down_proj(mul(self.act(self.gate_proj(x)), self.up_proj(x)))


class LlamaAttention(nn.Module):
    """Multi-head attention with optional GQA and RoPE."""

    def __init__(self, config: LlamaConfig) -> None:
        super().__init__()
        self.hidden_size = config.hidden_size
        self.n_heads = config.num_attention_heads
        self.n_kv_heads = config.num_key_value_heads
        self.head_dim = config.head_dim
        self.n_rep = self.n_heads // self.n_kv_heads
        self.use_gqa = self.n_rep > 1
        if self.use_gqa:
            self.q_weight = nn.Parameter(
                torch.empty(
                    self.hidden_size,
                    self.head_dim,
                    self.n_kv_heads,
                    self.n_rep,
                )
            )
            self.o_weight = nn.Parameter(
                torch.empty(
                    self.head_dim,
                    self.n_kv_heads,
                    self.n_rep,
                    self.hidden_size,
                )
            )
        else:
            self.q_weight = nn.Parameter(
                torch.empty(self.hidden_size, self.head_dim, self.n_heads)
            )
            self.o_weight = nn.Parameter(
                torch.empty(self.head_dim, self.n_heads, self.hidden_size)
            )
        self.k_weight = nn.Parameter(
            torch.empty(self.hidden_size, self.head_dim, self.n_kv_heads)
        )
        self.v_weight = nn.Parameter(
            torch.empty(self.hidden_size, self.head_dim, self.n_kv_heads)
        )
        self.reset_parameters()

    def reset_parameters(self) -> None:
        for p in (
            self.q_weight,
            self.k_weight,
            self.v_weight,
            self.o_weight,
        ):
            nn.init.normal_(p, std=0.02)

    def _project_q(self, x: Tensor) -> Tensor:
        q = gemm(x, self.q_weight, ndim=1, batch_ndim=0)
        return nntile_model_transpose(q, 2 if self.use_gqa else 1)

    def _project_kv(self, x: Tensor, weight: Tensor) -> Tensor:
        out = gemm(x, weight, ndim=1, batch_ndim=0)
        return nntile_model_transpose(out, 1)

    def _output(self, attn_out: Tensor) -> Tensor:
        attn_t = nntile_model_transpose(attn_out, 3)
        out_ndim = 3 if self.use_gqa else 2
        return gemm(attn_t, self.o_weight, ndim=out_ndim, batch_ndim=0)

    def _apply_rope(self, x: Tensor, sin: Tensor, cos: Tensor) -> Tensor:
        # sin/cos are ``[S, head_dim // 2]``. Extra leading modes of
        # ``x`` (heads, GQA n_rep, batch) are the kernel batch ``n``.
        if sin.device != x.device:
            sin = sin.to(x.device)
            cos = cos.to(x.device)
        return rope(sin, cos, x)

    def forward(
        self,
        x: Tensor,
        sin: Tensor | None = None,
        cos: Tensor | None = None,
        attn_mask: Tensor | None = None,
        *,
        is_causal: bool = True,
    ) -> Tensor:
        s = int(x.size(1))
        q = self._project_q(x)
        k = self._project_kv(x, self.k_weight)
        v = self._project_kv(x, self.v_weight)
        if sin is not None and cos is not None:
            q = self._apply_rope(q, sin, cos)
            k = self._apply_rope(k, sin, cos)
        batch_ndim = 2
        if self.use_gqa:
            k = _repeat_kv(k, self.n_rep)
            v = _repeat_kv(v, self.n_rep)
            batch_ndim = 3
        mask = attn_mask
        if mask is None and is_causal:
            mask = prepare_sdpa_mask(
                None,
                x,
                q_len=s,
                is_causal=True,
            )
        out = sdpa_kernel(
            q,
            k,
            v,
            mask=mask,
            batch_ndim=batch_ndim,
        )
        return self._output(out)


class LlamaDecoder(nn.Module):
    def __init__(self, config: LlamaConfig) -> None:
        super().__init__()
        self.input_layernorm = LlamaRMSNorm(
            config.hidden_size, eps=config.rms_norm_eps
        )
        self.self_attn = LlamaAttention(config)
        self.post_attention_layernorm = LlamaRMSNorm(
            config.hidden_size, eps=config.rms_norm_eps
        )
        self.mlp = LlamaMLP(config)

    def forward(
        self,
        x: Tensor,
        sin: Tensor | None = None,
        cos: Tensor | None = None,
        attn_mask: Tensor | None = None,
        *,
        is_causal: bool = True,
    ) -> Tensor:
        residual = x
        x = self.input_layernorm(x)
        x = self.self_attn(x, sin, cos, attn_mask, is_causal=is_causal)
        x = add(residual, x)
        residual = x
        x = self.post_attention_layernorm(x)
        x = self.mlp(x)
        return add(residual, x)


class LlamaModel(nn.Module):
    def __init__(self, config: LlamaConfig) -> None:
        super().__init__()
        self.config = config
        self.embed_tokens = Embedding(config.vocab_size, config.hidden_size)
        self.layers = nn.ModuleList(
            [LlamaDecoder(config) for _ in range(config.num_hidden_layers)]
        )
        self.norm = LlamaRMSNorm(config.hidden_size, eps=config.rms_norm_eps)
        # Host-built RoPE / causal-mask tables (NNGraph bind_data).
        # RoPE is ``[seq, head_dim // 2]``; batch is kernel n, not a table.
        self._rope_cache: dict[int, tuple[Tensor, Tensor]] = {}
        self._causal_mask_cache: dict[int, Tensor] = {}

    def _cached_rope(
        self, seq: int, device: torch.device
    ) -> tuple[Tensor, Tensor]:
        """Return ``[seq, head_dim // 2]`` sin/cos (built once, then reused).

        Built from ``arange(seq)``. The RoPE kernel applies that table
        across every batch row.
        """
        cached = self._rope_cache.get(seq)
        if cached is not None and cached[0].device == device:
            return cached
        pos_host = torch.arange(seq, dtype=torch.long, device="cpu")
        sin, cos = rope_sin_cos_from_position_ids(
            pos_host,
            self.config.head_dim,
            rope_theta=self.config.rope_theta,
        )
        if device.type != "cpu":
            sin = sin.to(device)
            cos = cos.to(device)
        self._rope_cache[seq] = (sin, cos)
        return sin, cos

    def _cached_causal_mask(
        self, seq: int, device: torch.device
    ) -> Tensor:
        """Return ``[seq, seq]`` BOOL causal mask (cached, like GPT-2)."""
        cached = self._causal_mask_cache.get(seq)
        if cached is not None and cached.device == device:
            return cached
        causal_mask = make_causal_sdpa_mask(seq, device=device)
        self._causal_mask_cache[seq] = causal_mask
        return causal_mask

    def clear_sequence_caches(self) -> None:
        self._rope_cache.clear()
        self._causal_mask_cache.clear()

    def warm_sequence_caches(
        self,
        *,
        batch_sizes: list[int] | tuple[int, ...] | None = None,
        seq_len: int,
        device: torch.device | str,
    ) -> None:
        """Prepare RoPE / causal-mask tables once for training reuse."""
        del batch_sizes
        device = torch.device(device)
        if seq_len < 1:
            raise ValueError(f"seq_len must be >= 1, got {seq_len}")
        self._cached_rope(seq_len, device)
        self._cached_causal_mask(seq_len, device)

    def forward(
        self,
        input_ids: Tensor,
        position_ids: Tensor | None = None,
        sin: Tensor | None = None,
        cos: Tensor | None = None,
        attn_mask: Tensor | None = None,
        *,
        is_causal: bool = True,
    ) -> Tensor:
        # Optional ``position_ids`` is HF-shaped leftover. Training is
        # ``arange(seq)`` for every batch row; B is consumed by RoPE.
        seq = int(
            position_ids.size(-1)
            if position_ids is not None
            else input_ids.size(-1)
        )
        if sin is None or cos is None:
            sin, cos = self._cached_rope(seq, input_ids.device)
        if attn_mask is None and is_causal:
            attn_mask = self._cached_causal_mask(seq, input_ids.device)
        x = self.embed_tokens(input_ids)
        for layer in self.layers:
            x = layer(x, sin, cos, attn_mask, is_causal=is_causal)
        return self.norm(x)


class LlamaCausal(nn.Module):
    """LlamaModel + lm_head (``nntile::model::llama::LlamaCausal``)."""

    def __init__(self, config: LlamaConfig) -> None:
        super().__init__()
        config.validate()
        self.config = config
        self.model = LlamaModel(config)
        self.lm_head = NntileLinear(
            config.hidden_size, config.vocab_size, bias=False
        )
        # Weight tying intentionally unsupported (independent lm_head).

    def forward(
        self,
        input_ids: Tensor,
        position_ids: Tensor | None = None,
        sin: Tensor | None = None,
        cos: Tensor | None = None,
        attn_mask: Tensor | None = None,
    ) -> Tensor:
        hidden = self.model(
            input_ids,
            position_ids=position_ids,
            sin=sin,
            cos=cos,
            attn_mask=attn_mask,
            is_causal=True,
        )
        return self.lm_head(hidden)


__all__ = [
    "LlamaAttention",
    "LlamaCausal",
    "LlamaConfig",
    "LlamaDecoder",
    "LlamaMLP",
    "LlamaModel",
    "LlamaRMSNorm",
    "make_causal_sdpa_mask",
]
